Making sure BatchExecutor doesn't hang when a connection is closed.

sduskis opened the pull request:
If a connection is closed, async executor will throw an exception, but only after it registered in the HeapSizeManager; the operation doesn't unregister in that case.  That leads to the client hanging forever.  Returning a future that immediately fails solves this problem.  The failure informs the FutureCallable from the HeapSizeManager that an error occured and that the operation can be marked as complete.
This fixes a problem a user found in #638 .

kevinsi4508 commented on bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/TestBatch.java:
```diff
@@ -366,4 +375,12 @@ public void testBatchGets() throws Exception {
+    Connection closedConnection = ConnectionFactory.createConnection(IntegrationTests.getConfiguration());
+    Table table = closedConnection.getTable(TABLE_NAME);
+    closedConnection.close();
+    thrown.expect(RetriesExhaustedWithDetailsException.class);
```
Can we verify the cause of the exception? connection closed related.

sduskis replied:
I guess so.  The point here is that the exception is thrown and the put doesn't hang forever.

kevinsi4508 replied:
The reason I prefer checking the cause (even it is IllegalStateException) is that we set up a precondition that will eventually force the code to throw some exception, we want to to verify the exception is thrown from those async calls not somewhere else.

For example:

batch{
throw e1; // unexpected exception that could be caught in our test code.
try {
  async  aclls
} catch {
  throw e2; // expected exception to be caught in our test code.
}

} // batch

If you believe checking IllegalStateException could be confusing, it is something we have fix, not a reason for us not to check it.

sduskis replied:
The reason not to check for a specific type of exception HERE is that it's not part of this specific test.  If we want to observe specific types of behavior for .close() exceptions, I would ask that we do it somewhere else.  This test is specifically made to test .put() batch operation not specific .close() behavior.

sduskis replied:
If the specifics of the exception on .close() change, it should not require changes of Batch testing, which is what we're testing here.  Getting into the specific details of the .close() exception here would require future changes.  I'd like to avoid those future changes and test specific behavior in unit tests, if that's ok.

sduskis replied:
Of note, we'd have to make sure that both Cloud Bigtable and HBase both show the same underlying behavior.

kevinsi4508 commented:
The test part is only concern, otherwise LGTMed.

kevinsi4508 commented:
Let's make it a nit. In general, I want to make sure the exception is the exact exception was set up, but not absolutely needed.

LGTM

sduskis commented:
It turns out that the test wasn't actually executing before.  I had to add @Test to make it run.  The test currently fails, so I'm going to fix that.  This will require another review once that's done.

sduskis commented on bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BatchExecutor.java:
```diff
@@ -155,33 +156,47 @@ public BatchExecutor(
         new RpcResultFutureCallback<T>(row, callback, index, results, resultFuture);
-    Futures.addCallback(future, futureCallback, service);
+    results[index] = null;
+    if (service.isShutdown()) {
```
Kevin, can you please checkout this change?   This is where the failure will happen.

kevinsi4508 replied:
This looks good.
Is there a unit test where we test this behavior?

if (closed)
  expect ConnectionClosedException
else
  expect async all

sduskis commented:
I added a unit test, and updated the integration test with more details.  I got rid of ConnectionClosedException because it's not an HBase Exception; HBase uses an IOException, so I think it's appropriate that we do as well.

sduskis commented:
I think that things look a lot better now with the additional testing.

kevinsi4508 commented:
LGTM
